[ZEPPELIN-4978] Build separate Zeppelin server and interpreter images

Reamer opened the pull request:
### What is this PR for?
This PR provides new Dockerfiles to build a Zeppelin server and interpreter image. This PR tries to solve https://github.com/apache/zeppelin/pull/3769#discussion_r434610135
I agree with @alexott that multiple interpreter images are difficult to handle. Currently neither the docker launcher nor the K8s launcher support switching to other interpreter images.

Some motivations:
 - conda is used to install all needed python and R packages 
   - conda is able to manage a whole environment and both languages
 - files are used for installing python and R dependency
   - it's easy to customize the content of files. No need to change the Dockerfile itself.
   - a file has more readability
   - reduce dependency calculation
 - a zeppelin-distribution image is needed
   - you need to build Zeppelin only once, if you want to build new images in parallel on a build cluster

### What type of PR is it?
Improvement

### Todos
* [ ] - Task

### What is the Jira issue?
* https://jira.apache.org/jira/browse/ZEPPELIN-4978

### How should this be tested?
* Build the images

### Questions:
* Does the licenses files need update? No
* Is there breaking changes for older versions? No
* Does this needs documentation? No

alexott commented:
Philipp - can we also solve the problem with access rights for Conda posted by some users on SO: https://stackoverflow.com/questions/62770514/zeppelin-anaconda-current-user-does-not-have-write-permissions, https://stackoverflow.com/questions/62828983/zeppelin-python-interpreter-virtual-environment-activation-is-failing - I've asked them to file JIRA, but it looks like that they didn't

Reamer commented:
> Philipp - can we also solve the problem with access rights for Conda posted by some users on SO: https://stackoverflow.com/questions/62770514/zeppelin-anaconda-current-user-does-not-have-write-permissions, https://stackoverflow.com/questions/62828983/zeppelin-python-interpreter-virtual-environment-activation-is-failing - I've asked them to file JIRA, but it looks like that they didn't

Docker containers should not be used like a virtual machine. The creation of virtual envs (if necessary at all) or the installation of other Python packages should be done when creating the interpreter image.
We should not fix it, because it works as intended.

If you need write access to a file system location within your container, you can use `-v' (docker) or volumes (K8s).

alexott commented:
We need to think about target audience in this case. If we're targeting analysts, they are mostly not the people who are creating Docker images, etc. Requirement to build custom image for every additional library that is a big barrier for adoption from my point of view. If we want to get people onto Zeppelin on K8S, we need at least provide very detailed documentation on how to configure volumes, etc., so they could install the packages that they need. 

I'm not very familiar with Conda, but I suspect that when you do `conda install` it will try to install packages to place where it's installed itself, and will break.

Reamer commented:
With the following change, every user should be able to modify Conda packages. However, I do not recommend this, because it can have side effects if several users are running on the same interpreter process.
```
    # Allow to modify conda packages
    chmod -R ug+rwX /opt/conda
```

zhoulii commented on k8s/zeppelin-server.yaml:
```diff
@@ -29,7 +29,7 @@ data:
   ZEPPELIN_K8S_SPARK_CONTAINER_IMAGE: spark:2.4.5
-  ZEPPELIN_K8S_CONTAINER_IMAGE: apache/zeppelin:0.9.0-SNAPSHOT
+  ZEPPELIN_K8S_CONTAINER_IMAGE: apache/zeppelin-interpreter:0.9.0-SNAPSHOT
   ZEPPELIN_HOME: /zeppelin
```
based on Dockerfiles(`scripts/docker/zeppelin-interpreter/Dockerfile`、`scripts/docker/zeppelin-server/Dockerfile`), value of `ZEPPELIN_HOME` should be `/opt/zeppelin`

zhoulii commented on scripts/docker/zeppelin-interpreter/Dockerfile:
```diff
@@ -0,0 +1,80 @@
+### COPY --from=zeppelin-distribution /opt/zeppelin/interpreter/zeppelin-interpreter-shaded-${VERSION}.jar ${Z_HOME}/interpreter/zeppelin-interpreter-shaded-${VERSION}.jar
+#### Copy specific interpreters,  replace "${interpreter_name}" with your interpreter. Of course you can repeat the line with defferent interpreter
+### COPY --from=zeppelin-distribution /opt/zeppelin/interpreter/${interpreter_name}  ${Z_HOME}/interpreter/${interpreter_name}
+
```
we should create directory `${Z_HOME}/logs` first or set USER to root, otherwise  when starting interpreter-server, we would meet this error:

`Container ENTRYPOINT failed to add passwd entry for anonymous UID
Log dir doesn't exist, create /opt/zeppelin/logs
mkdir: cannot create directory '/opt/zeppelin/logs': Permission denied`

Reamer replied:
Done in https://github.com/apache/zeppelin/pull/3859/files#diff-816be6e430bff3524979b549f7952db7R76

zhoulii commented on scripts/docker/zeppelin-interpreter/Dockerfile:
```diff
@@ -0,0 +1,80 @@
+# Decide:
+## 1) Build a huge image with all interpreters (default)
+COPY --from=zeppelin-distribution /opt/zeppelin/interpreter ${Z_HOME}/interpreter
+## 2) Build an image with only some interpreters
```
we should create directory `${Z_HOME}/run` in this Dockerfile or in `interpreter.sh`. 

 `${Z_HOME}/run` is pid directory, when starting interpreter-server, the pid file would write to this dir. However, `interpreter.sh` does not create this directory, so if not create first, we would meet the following error:

`/opt/zeppelin/bin/interpreter.sh: line 305: /opt/zeppelin/run/zeppelin-interpreter-md-shared_process--md-dubwak-12320.pid: No such file or directory`

Reamer replied:
Done in https://github.com/apache/zeppelin/pull/3859/files#diff-816be6e430bff3524979b549f7952db7R76

zhoulii commented:
Hi @Reamer , thanks a lot for creating this PR. I have build these three images and test them on a real k8s environment,  the comments left are about the problems I met.

zhoulii commented:
there is another problem,  when starting interpreter-server, [K8sRemoteInterpreterProcess#L283](https://github.com/apache/zeppelin/blob/master/zeppelin-plugins/launcher/k8s-standard/src/main/java/org/apache/zeppelin/interpreter/launcher/K8sRemoteInterpreterProcess.java#L283) would set `ZEPPELIN_HOME` in `interpreter pod` to `/zeppelin`, but the real `ZEPPELIN_HOME` in `interpreter pod` is `/opt/zeppelin`.

I think of two ways to slove this issue:
**solution 1:**  simply hard code `ZEPPELIN_HOME` to `/opt/zeppelin` in [Line 283](https://github.com/apache/zeppelin/blob/master/zeppelin-plugins/launcher/k8s-standard/src/main/java/org/apache/zeppelin/interpreter/launcher/K8sRemoteInterpreterProcess.java#L283)

**solution 2:** read `ZEPPELIN_HOME` from env variables. 
- first, add `ZEPPELIN_HOME` as ENV to `zeppelin-server` container in [zeppelin-server.yaml](https://github.com/apache/zeppelin/blob/master/k8s/zeppelin-server.yaml#L132)
`- name: ZEPPELIN_HOME
          valueFrom:
            configMapKeyRef:
              name: zeppelin-server-conf-map
              key: ZEPPELIN_HOME`
- then, change [Line 283](https://github.com/apache/zeppelin/blob/master/zeppelin-plugins/launcher/k8s-standard/src/main/java/org/apache/zeppelin/interpreter/launcher/K8sRemoteInterpreterProcess.java#L283) to
`envs.put("ZEPPELIN_HOME", envs.getOrDefault("ZEPPELIN_HOME", System.getenv("ZEPPELIN_HOME")));`

Reamer commented:
Hello @zhoulii,
Thank you very much for checking and reviewing this PR. I have changed ZEPPELIN_HOME in our example [configMapKeyRef](https://github.com/apache/zeppelin/blob/master/k8s/zeppelin-server.yaml#L33) to /opt/zeppelin. I have also added your provided code in K8sRemoteInterpreterProcess.java

I will change the default home of Zeppelin in a separate PR to conform to the [Filesystem Hierarchy Standard] (https://en.wikipedia.org/wiki/Filesystem_Hierarchy_Standard).

zhoulii commented:
> Hello @zhoulii,
> Thank you very much for checking and reviewing this PR. I have changed ZEPPELIN_HOME in our example [configMapKeyRef](https://github.com/apache/zeppelin/blob/master/k8s/zeppelin-server.yaml#L33) to /opt/zeppelin. I have also added your provided code in K8sRemoteInterpreterProcess.java
> 
> I will change the default home of Zeppelin in a separate PR to conform to the [Filesystem Hierarchy Standard] (https://en.wikipedia.org/wiki/Filesystem_Hierarchy_Standard).

LGTM , Great job.
